assets/semgrep_rules/c/chromium-bind-uaf.yaml

```diff
@@ -8,7 +8,7 @@ rules:
         - pattern: $BASE::BindRepeating($...ARGS)
         - pattern: BindRepeating($...ARGS)
       - metavariable-comparison:
-          comparison: not re.match("base::Unretained\(this\)", str($...ARGS)) and re.match("base::Unretained", str($...ARGS))
+          comparison: re.match("base::Unretained", str($...ARGS))
```

Comment:
There's also another case when things can go wrong.

Let's implement this as two separate rules:
1. Catch all `base::Unretained(...)` calls.
2. Catch `base::Bind*` calls where:
* first argument is not a class function pointer (i.e. doesn't doesn't contain `::`)
* second argument is `this`. Example:

```
// this is bad and most likely will explode:
auto func = [](MyClass* c, int val) {
  if (val > 10) {
    c->DoSomething();
  } else {
    c->DoSomethingElse();
  }
};
return base::BindOnce(&func, this);

// this is fine:
return base::BindOnce(&MyClass::DoSomethingConditionally, weak_ptr_factory_.GetWeakPtr());
```

Reply:
I created an issue to track; it will take a bit more time since I also need to adapt the test cases.
Will sort it out later today!

https://github.com/brave/security-action/issues/222